### ISSUE TITLE
Upload the weight's diff if it fails to push for some reason

### ISSUE THREAD

**Issue by joao-paulo-parity** [paritytech/bench-bot#54]
At the moment, in case some error occurs while the bot is trying to push the generated weights (e.g. https://github.com/paritytech/substrate/pull/9686#issuecomment-915911689), the bot will simply error. It would be better to upload the weight's diff that the effort isn't lost (some benchmarks take a long time to run).

[Github limits the amount of characters per-comment](https://github.community/t/maximum-length-for-the-comment-body-in-issues-and-pr/148867/2) so that's something to also take into account before posting the diff.

**Comment by athei** [comment]
You could use `git format-patch` to create a patch file and compress this and use a file upload instead of posting it as text. This would circumvent the limitation and also makes it easier to apply.

**Comment by shawntabrizi** [comment]
CC @ggwpez

We were already thinking of generating raw JSON output files from all the benchmarks so that we could generate visual representations of the benchmarks in case people want to look closer at the results.

From this, we should also be able to easily generate the weight files if they could not be committed.

**Comment by ggwpez** [comment]
So where do we upload it to? I don't think creating Gists is optimal.  
Maybe some S3 compatible server to throw the JSON files in?

**Comment by shawntabrizi** [comment]
Can we not just leave it on the benchmarking machine?

I guess we would be concerned about serving files while it was benchmarking?

If we need another server that works and can be allocated, otherwise, gist is probably not the worst idea.

**Comment by ggwpez** [comment]
> Can we not just leave it on the benchmarking machine?
> I guess we would be concerned about serving files while it was benchmarking?

Yes, just running a webserver there *could* interfere with the results.  
In the best case we would make that data publicly available, so that other people can also analyze Substrate+Polkadot weights.  

> If we need another server that works and can be allocated, otherwise, gist is probably not the worst idea.

If we just throw them in single Gist files we will not be able to index/order or find them ever again.  
To not go overboard for the first version of this; some kind of indexable file storage e.g. S3-compatible Bucket storage would be fine I think.

**Comment by joao-paulo-parity** [comment]
I'm assigning myself to this task since I plan to work on it sometime soon

**Comment by athei** [comment]
> So where do we upload it to? I don't think creating Gists is optimal. Maybe some S3 compatible server to throw the JSON files in?

I attachments to postings I make in github by dragging them in here and it will just upload them "somewhere". Can't the bot do the same thing?

**Comment by ggwpez** [comment]
> I attachments to postings I make in github by dragging them in here and it will just upload them "somewhere". Can't the bot do the same thing?

Yes for the diff that is enough. Sorry for mixing up two issues, my concern was regarding the raw output.

**Comment by koenw** [comment]
> If we just throw them in single Gist files we will not be able to index/order or find them ever again.

We would need to keep an index of Gists, perhaps in a file (or files) that we can commit/push.

Is there any other downside to Gists? It seems simpler to me than using an additional service. 

We would also easily be able to generate (or it could just be) a JSON of all benchmarks like @shawntabrizi mentioned.

**Comment by joao-paulo-parity** [comment]
Recent incidents required manual intervention:

- https://github.com/paritytech/substrate/pull/10617#issuecomment-1028927709
- https://github.com/paritytech/substrate/pull/10617#issuecomment-1019885368
